server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -522,150 +529,229 @@ public IndexMetaDataGenerations indexMetaDataGenerations() {
      *                           from cached bytes that we trust to not contain broken generations.
      */
     public static RepositoryData snapshotsFromXContent(XContentParser parser, long genId, boolean fixBrokenShardGens) throws IOException {
+        if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+            throw new ElasticsearchParseException("start object expected");
+        }
+
         final Map<String, SnapshotId> snapshots = new HashMap<>();
         final Map<String, SnapshotState> snapshotStates = new HashMap<>();
         final Map<String, Version> snapshotVersions = new HashMap<>();
         final Map<IndexId, List<SnapshotId>> indexSnapshots = new HashMap<>();
+        final Map<String, IndexId> indexLookup = new HashMap<>();
```

Comment:
Fill this up during parsing instead of doing another loop at the end of this method over all the indices